hal/device.c

```diff
@@ -0,0 +1,49 @@
+#include "device.h"
+
+static struct device_type *devices;
+
+static struct device_type **find_device(const char *name, unsigned len) {
+  struct device_type **p;
+  for (p = &devices; *p; p = &(*p)->next)
+    if (strlen((*p)->name) == len && strncmp((*p)->name, name, len) == 0) break;
+  return p;
+}
+
+device_t *device(const char *type) {
+  struct device_type **p;
+  if (devices->next) {
+    fprintf(stderr, "No device type registered!");
+    return NULL;
+  }
+  p = find_device(type, strlen(type));
+  if (*p) {
+    fprintf(stderr, "Device type %s not found", type);
```

Comment:
Shall we integrate logger here?

Reply:
I think we should focus on HAL subject here. Logger and and other logging message we would resolve by #34 .

Reply:
Add `TODO` and/or `FIXME`.